Rework priority groups, take 2

tomaka opened the pull request:
(this PR is the finished version of https://github.com/paritytech/substrate/pull/7374)
polkadot companion: https://github.com/paritytech/polkadot/pull/2095
close https://github.com/paritytech/substrate/issues/6087

cc @rphmeier @ordian @eskimor 
While you don't have to review the PR, I would appreciate opinions on the design described below.

This PR refactors the slots attribution system of the networking.

# Before this PR

A Substrate or Polkadot node establishes TCP connections with other libp2p-compatible nodes.
The set of connected peers is divided in two: the ones that have a slot, and the ones that don't.

The ones that don't have a slot are able to perform requests using request-response protocols, but are disconnected after 10 seconds of inactivity. Performing a Kademlia iterative query, for example, requires establishing connections to many nodes, and these connections don't need a slot.

The ones that have a slot are the ones notifications substreams are opened with. Multiple notifications protocols exist (block announces, transactions, grandpa, and, in Polkadot, collation and validity), and when a slot is attributed to a peer, one substream for each notifications protocol is opened.
As long as a substream is maintained using the block announces notification protocol (special-cased for this purpose), the peer keeps its slot, even if this peer doesn't support or has refused the protocols other than the block announces. As such, it is for example possible for all slots to be filled only with peers that only support block announces.

Due to backwards-compatibility issue, once a peer has performed (successfully or not) its handshake on all notifications protocols substreams, an `Event::NotificationStreamOpened` event is reported to the event subscribers. One such event is emitted for *all* notifications protocols, even if the substream of said protocol hasn't been accepted or isn't supported! This will lead the rest of the codebase to try send notifications using this protocol to a peer even though there is no substream for it.

Slots are attributed in the following way:

- 25 incoming slots, for peers that initiate opening the substreams to the local node.
- 25 outgoing slots, assigned randomly and based on reputation.
- The API of `sc-network` makes it possible to define a list of *priority groups* containing peers that automatically have a slot attributed to them.
- Reserved nodes, similar to priority groups, but defined by the end user with the `--reserved-node` CLI option.

Priority groups are currently used for two purposes: the authority-discovery module maintains a priority group containing 10 randomly-chosen validators, and Polkadot maintains a priority group containing the validators assigned to the current parachain.

When the user passes the `--reserved-only` CLI flag, all slots are closed except for nodes passed through `--reserved-node`.

## Problems with the current situation

The problems with the current situation are:

- We need to guarantee connectivity between collators and validators. Validators don't know in advance who is a collator and who isn't, and it is possible, at the moment, for a validator to have all its incoming slots filled with non-collators, in which case collators will be unable to collate.
- It is not desirable to make the code base believe that notifications protocols substreams are open, even when it is not the case.
- It is not desirable to open the collation notifications protocol substream with non-validators/non-collators, and it is not desirable to open the validity notifications protocol substream with non-validators.

# After this PR

The networking is now using *peer sets*. A peer set is associated to each notifications protocol. As such, in a Polkadot node, there exists five peer sets: block announces, transactions, grandpa, collation, validity.

Rather than being global, slots are now specific to a peer set. Each TCP connection can have, or not, a slot open with regards to each peer set. For example, a connection can have a slot for block announces but not for grandpa.

The way slots are attributed depends on the set.

The block announces peer set keeps the same slots attribution mechanism as the previously-global slots-attribution mechanism: 25 incoming, 25 outgoing, and reserved nodes. Priority groups have been removed from the codebase.

The transactions and grandpa slots attribution is configured to be permanently in "reserved only" mode, meaning that only a specific list of nodes are allowed to have a slot. This list of nodes is constantly kept synchronized with the nodes that occupy a block announces slot.
In other words, the transactions and grandpa notifications substreams keep the same behaviour as before this PR: they follow the block announces slots. The one exception is that no `Event::NotificationStreamOpened` event is reported anymore if there isn't actually a substream open.

Similarly, the validity peers sets slots attribution is configured to be permanently in "reserved only" mode, and its list of reserved nodes is meant to follow the list of validators discovered by the authority-discovery.
The mechanism where the authority-discovery module maintains a priority group of 10 randomly-chosen validators has been removed.

The collation peers set slots attribution is configured to have 25 inbound slots. Its list of reserved nodes is meant to follow the list of validators assigned to a specific parachain. As explained above, the links established for collation are asymmetrical: the collators know which validators to be connected to, and uses reserved nodes for that purpose, but validators don't know who the collators are, hence the 25 inbound slots.

Please let me know whether these rules regarding the validity and collation peer sets make sense.

These five peer sets are how the `generic_proto/behaviour.rs` and `generic_proto/handler.rs` files function, plus everything in `sc_peerset`.
However, the API of `sc_network` exposes the concepts of "default set" and "extra sets".
The default sets refers to the block announces and transactions, which are handled internally, and the extra sets refers to the peers sets/notifications protocols registered by API users.
Two events have been added: `SyncConnected` and `SyncDisconnected`, to report to users slots being attributed on the default set. This is necessary in order to implement the grandpa reserved nodes tracking of the block announces slots.

The `protocol.rs` files acts as the translation layer, as it is the file that handles the block announces and transactions in its internals. The `behaviour.rs` and `service.rs` files are only aware of the concepts of default set and extra sets, and expose APIs where extra sets are referred to by their notifications protocol names.
The `--reserved-nodes` and `--reserved-only` CLI options only refer to the default set.

# Concerns

## Additional round-trip

A round-trip is added when it comes to opening transactions and grandpa substreams. Previously, all substreams would immediately open simultaneously. Now, the block announces slot is first attributed, and it is only once the block announces handshake has been performed that transaction and grandpa slots are attributed to the peer.
I don't think that this really has big consequences, as connections for block-announces/transactions/grandpa are mostly stable and are used only for gossiping. But I admit that I did this in order to make the implementation easier to write, and to not expose complicated APIs. If that turns out to be a problem, it should be possible to expose APIs that instruct the implementation to "group" multiple protocols together.

## --reserved-only might be confusing

The `--reserved-only` flag now only acts on the default peers set (block announces, transactions, and, by extension, grandpa).
This means that validators and nodes configured for collating will still establish connections and substreams and gossip messages to non-reserved nodes.

This behaviour could cause some confusions. However, the alternative, which would be to apply `--reserved-only` to everything, would cause the networking to not work.

## Might contain bugs

This is a big PR that refactors a lot of code, and it is likely that bugs have been introduced. It should be thoroughly tested before reviewing, especially considering that I'm been lenient on the reviewing of https://github.com/paritytech/substrate/pull/7464.

tomaka commented:
Testing is showing that, when connecting to Polkadot, we don't receive any notification on sets 1 and 2 (notifications and grandpa).

I believe that there is a compatibility problem with the fact that we only later add peers to these sets.
When the remote tries to open their transaction and grandpa substreams, they face a rejection from the local node. Later, when the local node attempts to connect, the remote still remembers this rejection and doesn't try to re-open its own substream.

mxinden commented on client/network/src/service.rs:
```diff
@@ -942,87 +910,129 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
-	// future-proofing, see https://github.com/paritytech/substrate/pull/7247#discussion_r502263451.
-	pub async fn add_to_priority_group(&self, group_id: String, peers: HashSet<Multiaddr>) -> Result<(), String> {
+	// NOTE: technically, this function only needs `Vec<PeerId>`, but we use `Multiaddr` here for convenience.
+	pub fn remove_peers_set_reserved(&self, protocol: Cow<'static, str>, peers: HashSet<Multiaddr>) -> Result<(), String> {
```
```suggestion
	pub fn remove_peers_from_reserved_set(&self, protocol: Cow<'static, str>, peers: HashSet<Multiaddr>) -> Result<(), String> {
```

Would be easier for me to read.

mxinden replied:
```suggestion
	pub async fn remove_peers_set_reserved(&self, protocol: Cow<'static, str>, peers: HashSet<Multiaddr>) -> Result<(), String> {
```

See reasoning above on `pub fn add_peers_set_reserved`.

mxinden commented on client/peerset/src/lib.rs:
```diff
@@ -35,22 +48,40 @@ pub use libp2p::PeerId;
+/// Can be constructed using the `From<usize>` trait implementation based on the index of the set
+/// within [`PeersetConfig::sets`]. For example, the first element of [`PeersetConfig::sets`] is
+/// later referred to with `SetId::from(0)`. It is intended that the code responsible for building
+/// the [`PeersetConfig`] is also responsible for constructive the [`SetId`]s.
```
```suggestion
/// the [`PeersetConfig`] is also responsible for constructing the [`SetId`]s.
```

mxinden commented on client/peerset/src/lib.rs:
```diff
@@ -190,11 +232,10 @@ pub struct PeersetConfig {
+	/// For each set, lists of nodes that don't occupy slots and that we should try to always be
+	/// connected to, and whether only reserved nodes are accepted. Is kept in sync with the list
+	/// of non-slot-occupying nodes in [`Peerset::data`].
+	reserved_nodes: Vec<(HashSet<PeerId>, bool)>,
```
Instead of treating the `SetId` as an index into the `Vec`, why not use a `HashMap` instead? I would expect this to be less error prone, especially since anyone could create their own via `SetId::from(_)`.
```suggestion
	reserved_nodes: HashMap<SetId, (HashSet<PeerId>, bool)>,
```

tomaka replied:
I don't really understand how that is better.
The valid values for `SetId` range from `0` to `config.sets.len()`. If we put a `HashMap` instead, well, the valid values would remain the same.

Note that managing `SetId`s is entirely scoped to `protocol.rs`. Nothing outside of `protocol.rs` is supposed to generate a `SetId`, apart from `behaviour.rs` that converts them back and forth with a `usize`.

mxinden replied:
Both approaches seem a bit brittle off the top of my head, e.g. for the `Vec` approach with `assert!(set < self.sets.len());`, but also for the `HashMap` approach with the `unwrap` calls needed if an entry doesn't exist.

Whether `HashMap` is better than a `Vec` is quite subjective. I am fine with this given that `SetId` is not exposed outside of `client/network` and given the recent changes through https://github.com/paritytech/substrate/pull/7700/commits/ebaeaab210fdea2f66ae217ee7b8d2f57e7ea38c moving the first two set ids to constants.

mxinden commented on client/peerset/src/lib.rs:
```diff
@@ -216,159 +257,178 @@ impl Peerset {
+	/// Adds a node to the given set. The peerset will, if possible and not already the case,
+	/// try to connect to it.
+	///
+	/// > **Note**: This as the same effect as [`PeersetHandle::add_to_peers_set`].
```
```suggestion
	/// > **Note**: This has the same effect as [`PeersetHandle::add_to_peers_set`].
```

mxinden commented on client/peerset/src/lib.rs:
```diff
@@ -216,159 +257,178 @@ impl Peerset {
+
+		if reserved_only {
+			// Disconnect all the nodes that aren't reserved.
+			for set_index in 0..self.data.num_sets() {
```
I am a bit confused by the fact that `on_set_reserved_only` is specific to a `SetId`, but still the logic below disconnects peers of all sets in case they are not reserved. Why is that?

tomaka replied:
Fixed

mxinden commented on client/peerset/src/lib.rs:
```diff
@@ -403,27 +463,35 @@ impl Peerset {
+
+				drop(peer_reputation);
+
+				// If the peer reaches a reputation of 0, and there is no connect to it,
```
```suggestion
				// If the peer reaches a reputation of 0, and there is no connection to it,
```

mxinden commented on client/network-gossip/src/lib.rs:
```diff
@@ -97,8 +103,26 @@ impl<B: BlockT, H: ExHashT> Network<B> for Arc<NetworkService<B, H>> {
+	fn add_set_reserved(&self, who: PeerId, protocol: Cow<'static, str>) {
+		let addr = iter::once(multiaddr::Protocol::P2p(who.into()))
+			.collect::<multiaddr::Multiaddr>();
+		let result = NetworkService::add_to_peers_set(self, protocol, iter::once(addr).collect());
```
```suggestion
		let result = NetworkService::add_peers_set_reservedself, protocol, iter::once(addr).collect());
```

Should the node not be added as a reserved node?

tomaka replied:
Indeed. I changed that here: https://github.com/paritytech/substrate/pull/7700#issuecomment-741768066
I'm a bit unsure what to do.

mxinden commented on client/network-gossip/src/lib.rs:
```diff
@@ -97,8 +103,26 @@ impl<B: BlockT, H: ExHashT> Network<B> for Arc<NetworkService<B, H>> {
+	fn remove_set_reserved(&self, who: PeerId, protocol: Cow<'static, str>) {
+		let addr = iter::once(multiaddr::Protocol::P2p(who.into()))
+			.collect::<multiaddr::Multiaddr>();
+		let result = NetworkService::remove_from_peers_set(self, protocol, iter::once(addr).collect());
```
Same here, should this not call `remove_peers_set_reserved`?

```suggestion
		let result = NetworkService::remove_peers_set_reserved(self, protocol, iter::once(addr).collect());
```

mxinden reviewed (commented):
I did a first pass over this pull request. I don't yet fully understand the bigger picture, thus some suggestions might be wrong due to misunderstandings.

tomaka commented on client/network/src/service.rs:
```diff
@@ -942,87 +910,129 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
-	// NOTE: even though this function is currently sync, it's marked as async for
-	// future-proofing, see https://github.com/paritytech/substrate/pull/7247#discussion_r502263451.
-	pub async fn set_priority_group(&self, group_id: String, peers: HashSet<Multiaddr>) -> Result<(), String> {
+	pub fn add_peers_set_reserved(&self, protocol: Cow<'static, str>, peers: HashSet<Multiaddr>) -> Result<(), String> {
```
This wasn't possible anymore due to the need to integrate them [here](https://github.com/paritytech/substrate/blob/0e91d3d5d340058726bf1230e69d755a793fd7cf/client/network-gossip/src/lib.rs#L106-L122).
Making this `Network` trait an async trait also isn't possible because it's called from [a `poll` function](https://github.com/paritytech/substrate/blob/0e91d3d5d340058726bf1230e69d755a793fd7cf/client/network-gossip/src/bridge.rs#L182). I didn't have the courage to rewrite this `Future` as an async function.

dvdplm replied:
I believe the correct method name here would be `add_peer_set_reserved`; "a peer set" –> "many peer sets".

tomaka commented on client/peerset/tests/fuzz.rs:
```diff
@@ -71,70 +78,101 @@ fn test_once() {
+				3 => {
+					if let Some(id) = connected_nodes.iter().choose(&mut rng).cloned() {
+						connected_nodes.remove(&id);
+						peerset.dropped(SetId::from(0), id);
```
This is a test, and these numbers are generated randomly [here](https://github.com/paritytech/substrate/blob/9c4b449c3273fd9086018f0d6927ff74e2d6d327/client%2Fpeerset%2Ftests%2Ffuzz.rs#L83-L86). I don't think using enums would make this in any way more simple. Additionally, this code was already there before this PR, and I'd prefer to not change this here.

mxinden commented on client/network-gossip/src/bridge.rs:
```diff
@@ -178,6 +178,12 @@ impl<B: BlockT> Future for GossipEngine<B> {
 				ForwardingState::Idle => {
 					match this.network_event_stream.poll_next_unpin(cx) {
 						Poll::Ready(Some(event)) => match event {
+							Event::SyncConnected { remote } => {
```
As far as I know `client/network-gossip` is only used in `client/finality-grandpa`. It is kept separate to potentially allow others to use it as well.

This change ties `client/network-gossip` directly to `client/finality-grandpa` as it enforces the users peer set to always equal the sync peer set.

I don't quite understand all changes in this pull request yet. In case we stick to this solution, would it not make sense to move `client/network-gossip` into `client/finality-grandpa` or at least call out that `client/network-gossip` is not supposed to be used by anyone other then `client/finality-grandpa`?

tomaka replied:
Indeed, `network-gossip` now always maintains a set of peers equal to the sync one. The difference is that this is now decided by `network-gossip`, while before this was enforced by `sc-network`.

I don't think this means that `network-gossip` is tied to `finality-grandpa`, only that it is tied to the "main chain" the node manages.

I thought about this a bit and went for the solution of adding some documentation in https://github.com/paritytech/substrate/pull/7700/commits/9e702d6c99c948c7fe2f29e54db0da77981ccb49.

rphmeier commented:
@tomaka 

Thanks. The high-level description of before-and-after is really helpful to understand the context. 

I have a few questions about the content, but it generally seems clear.

> The collation peers set slots attribution is configured to have 25 inbound slots. Its list of reserved nodes is meant to follow the list of validators assigned to a specific parachain. As explained above, the links established for collation are asymmetrical: the collators know which validators to be connected to, and uses reserved nodes for that purpose, but validators don't know who the collators are, hence the 25 inbound slots.

If I understand correctly, in the previous implementation, inbound slots were cleared when block announcements weren't made on the slot. What is the automated criterion for clearing the inbound slots for the collator protocol now? What input does the higher-level code have on this criterion ?

> These five peer sets are how the generic_proto/behaviour.rs and generic_proto/handler.rs files function, plus everything in sc_peerset.

Do you mean that the validation & collation peer-sets are actually defined in Substrate? That seems a little leaky, as it is a higher-level concern. Not something I would block this on, but worth knowing about. And if so, do you have any plan, however vague, to remove that from Substrate per-se?

> The transactions and grandpa slots attribution is configured to be permanently in "reserved only" mode, meaning that only a specific list of nodes are allowed to have a slot. This list of nodes is constantly kept synchronized with the nodes that occupy a block announces slot.

Testing my understanding: transactions and grandpa slots are kept synced with the block announce slots by assigning the reserved peers to be equal to whoever happens to be in the block announce slots at any given moment?

----

The last questions I have are about number of slots & dormant connections as we move towards 1000 validators. Is the number of slots cli-configurable or at least in network configuration? What guidelines should we use to determine the number of slots we use? And is there any notion of peers we have 'dormant' connections to so we can speak with them on-demand but potentially very intermittently without doing a handshake?

burdges commented:
I'm slightly confused by the "peer sets: block announces, transactions, grandpa, collation, validity" bit, but this  sounds like progress.  :)  

At a high level, we have three connection categories: validator-validator, validator-collator, and validator-others (in future hierarchical relay chains shall complicate this, but they're way way off).

A priori, we'd want all validator-validator connections open fairly permanently, so if there are 1000 validators then they all open direct connections to one another.  We should push low-level networking to support ala these connections, ala QUIC, node shutdown if ulimit set wrong, say validators cannot run on OSX or Windows, or anything else this requires.  We can relax this slightly like num_validators/3 connections or so, but anything like 25 kills our scalability and soundness.

We've three or four usages for validator-collator connections:  (a) collators informing validators about candidates aka collations, (b) collators fetching incoming XCMP messages from the sending parachain validators, (c) validators informing collators about relay chain's progress, and (d) validators providing parachains with censorship resistance.

(a) I'd suggest that collators open direct connections with their assigned validators when they have a collation ready.  If (n-k+1)-of-n validators reject this connection attempt then that parachain shall not make progress, where availability requires k backers out of n parachain validators.  This makes DoSing the parachain access to the relay chain easy, but that could be solved by adding flexibility in core usage.

(b) We've discussed [many XCMP transports](https://hackmd.io/@rgbPIkIdTwSICPuAq67Jbw/HJBrUKhuI) but only one pattern covers all cases: collators fetch incoming XCMP messages from the sending parachain validators, which requires short lived connections.  We'll explore other patterns in future if we noticed bottlenecks or whatever.

(c) We need collators to track the relay chain so that they can build correct state references for the relay parent they select for the candidate they build.  At present, collators should be full nodes of the relay chain, which dictates many other networking decisions.  We've numerous unrealized optimizations here, ala contextual execution and XCMP state handling, which should enable more direct and optimized networking choice eventually, but not today.  

(d) We do not provide parachains with censorship resistance currently because not all parachain designs benefit, so this does not fit within the MVP.  We should support this eventually however because we'll need this for our own internal efforts.  We'd grant collators with upcoming block production slots some preference for either fetching the block from an approval checker or doing a full reconstruction.

We need validator-others connections so that non-validators can become validators and relay chain transactions.

What does this all give us?  We've (1) semi-permenent complete graph on validators, (2) collators opening ephemeral connections to validators, both theirs (a) and random others (b), (3) validators asisting non-validator full nodes with tracking the chain for (c) and other usages, and (4) the future work of validators ensuring parachain block availability for anyone besides validators (d). 

We depend upon validator-validator connections to communicate approval checker assignments for parachain soundness, so any network level, DoS, etc. attack should impact validator-validator connections last. 

If I understand, we're currently doing (2) not using direct connections but instead using the mechanisms designed for (3), largely out of fear of demanding more connections of validators.  This if fine for now, but it should be reevaluated next year.  If we ever do nQUIC with noise IK then then validators could prove themselves to other validators inside one UDP packet, which helps making (1) sacrosanct.  Afaik, we cannot prove upcoming collator status in only one UDP packet.  We thus create a DoS vector with (2) but ideally it should only DoS (2).  I'm unsure about relative priorities of (2-4) but maybe (3) outranks (2) slightly.

eskimor commented:
Sounds sensible to me. With regards to `--reserved-only` .. is it possible for the user to make the network work by specifying reserved nodes on the command line for example? In this case it could be better to make it do what it says and to just add a warning in the man page, that you need to add nodes to the reserved set to make this work. Having a way of enforcing a particular set of nodes and no others, e.g. for testing sounds useful to me.

In that case, if we need the less strict behavior as well, we could introduce another more descriptive option for this.

On the other hand, if there is no way ever, somebody would want that more strict behavior, it should be fine to change it and just explain the option in the man page. Problem is, if that is not the case, we will end up eventually with a `--reserved-only-really` option and that would be a pity.

tomaka commented:
> If I understand correctly, in the previous implementation, inbound slots were cleared when block announcements weren't made on the slot. What is the automated criterion for clearing the inbound slots for the collator protocol now? What input does the higher-level code have on this criterion ?

To clarify: we don't detect situations where no block announcement message is being sent. Inbound slots are only cleared in case of disconnect, or if the reputation of the remote drops too low.
Similarly, inbound slots for collation are cleared automatically only in case of disconnect, of if the reputation drops too low.

In terms of higher-level code, calling `remove_from_peer_set` can be called in order to disconnect the node. This method has the effect that the `PeerId` is not longer considered as belonging to the given set, which means that the local node will also not attempt to connect back to this peer if any out slot was available.

> Do you mean that the validation & collation peer-sets are actually defined in Substrate? That seems a little leaky, as it is a higher-level concern. Not something I would block this on, but worth knowing about. And if so, do you have any plan, however vague, to remove that from Substrate per-se?

No, see [the Polkadot companion](https://github.com/paritytech/polkadot/pull/2095/files#diff-23273d09273311f937a0cc595e690191ae03d0c13dbf438728a53c1af0a3d285R88-R105).

The API surface that `sc-network` exposes is: a "default peer set" (managed internally and used for syncing, block announces and transactions), and "extra peer sets" passed as part of the configuration and managed entirely by higher-level code. See [this](https://github.com/paritytech/substrate/blob/c69bf7a1559a37462d39f5157d794a54989fe4db/client/network/src/config.rs#L375-L378).

> Testing my understanding: transactions and grandpa slots are kept synced with the block announce slots by assigning the reserved peers to be equal to whoever happens to be in the block announce slots at any given moment?

Yes. Transactions and grandpa slots are consequently in "reserved-only" mode.

tomaka commented:
> A priori, we'd want all validator-validator connections open fairly permanently, so if there are 1000 validators then they all open direct connections to one another. We should push low-level networking to support ala these connections, ala QUIC, node shutdown if ulimit set wrong, say validators cannot run on OSX or Windows, or anything else this requires. We can relax this slightly like num_validators/3 connections or so, but anything like 25 kills our scalability and soundness.

Yes, the idea is that Polkadot's code maintains the list of reserved node of the "validity" set to be equal to the list of on-chain validators (after being passed through the authority-discovery system). If there are 1000 on-chain validators, there will be 1000 entries in the reserved nodes, and the networking code will attempt to open a substream with all 1000.

> I'd suggest that collators open direct connections with their assigned validators when they have a collation ready.

At the moment, collators try to maintain a direct connection with their assigned validators at all time, even if no collation is ready. This is done in the same way: fetching the list of assigned validators from the runtime, then adding them to a list of reserved nodes.

> If I understand, we're currently doing (2) not using direct connections but instead using the mechanisms designed for (3), largely out of fear of demanding more connections of validators.

I don't really think that fear is justified. We're supposed to be rather easily able to maintain tens of thousands of TCP connections open.

> We thus create a DoS vector with (2) but ideally it should only DoS (2). I'm unsure about relative priorities of (2-4) but maybe (3) outranks (2) slightly.

To me, the major risk of DoS attack is to block the TCP listening port of a validator by opening a large number of connections until the limit of the node (which right now has been arbitrarily fixed at 10,000 incoming connections) is reached.
However, the node being attacked should still be able to establish outbound connections. To prevent validator A and B from connecting to each other, you'd need to attack both A and B at the same time.

Another risk is that, since we don't prioritize the bandwidth of important connections such as validator-validator connections, an attacker can send enough traffic to use up all the node's available bandwidth, and slow down the node enough that timings are missed.

tomaka commented:
> is it possible for the user to make the network work by specifying reserved nodes on the command line for example?

Yes

> In this case it could be better to make it do what it says and to just add a warning in the man page, that you need to add nodes to the reserved set to make this work. Having a way of enforcing a particular set of nodes and no others, e.g. for testing sounds useful to me.

The problem here is that collators *need* to be connected to validators, and validators need to be connected to other validators, for the Polkadot networking as a whole to work.
It would be really tedious to have to find out in advance who the hundreds of validators are, and this list changes dynamically after startup.

burdges commented:
> However, the node being attacked should still be able to establish outbound connections. To prevent validator A and B from connecting to each other, you'd need to attack both A and B at the same time.

Interesting.  There are DoS attacks at all levels, but if validator-validator connections staying live prevents most higher level DoS attacks then that's super handy.  :)

> Another risk is that, since we don't prioritize the bandwidth of important connections such as validator-validator connections, an attacker can send enough traffic to use up all the node's available bandwidth, and slow down the node enough that timings are missed.

We should "one day" think about prioritizing bandwidth too of course, especially so that validators could always send outgoing gossip messages.  

Important question:  Are routers elsewhere enforcing flow control and providing DoS protections for TCP?  If so, this could provides one advantage for TCP over QUIC, assuming we use this correctly.  We could consider QUIC among validators with TCP for validators-others too, although maybe even having UDP ports open creates risks.  Or maybe even UDP packets get dropped elsewhere in ways that help prevent UDP DoS attacks?  In that case, our problem could become "explaining" our QoS goals to the network elsewhere.

mxinden commented on client/network-gossip/src/lib.rs:
```diff
@@ -75,8 +75,14 @@ pub trait Network<B: BlockT> {
+	/// Adds the peer to the set of peers to be connected to with this protocol.
+	fn add_set_reserved(&self, who: PeerId, protocol: Cow<'static, str>);
+
+	/// Removes the peer to the set of peers to be connected to with this protocol.
```
```suggestion
	/// Removes the peer from the set of peers to be connected to with this protocol.
```

mxinden commented on client/network/src/behaviour.rs:
```diff
@@ -352,38 +328,36 @@ Behaviour<B, H> {
+				self.events.push_back(BehaviourOut::NotificationStreamReplaced {
+					remote,
+					protocol,
+					notifications_sink: notifications_sink.clone(),
```
```suggestion
					notifications_sink,
```

mxinden commented on client/network/src/protocol.rs:
```diff
@@ -632,7 +692,7 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
 					who
 				);
-				self.disconnect_peer(&who);
+				self.behaviour.disconnect_peer(&who, sc_peerset::SetId::from(0));
```
Just by reading `sc_peerset::SetId::from(0)` it is hard to know that one is disconnecting from the block request peer set. What do you think of defining a constant e.g. `BLOCK_ANNOUNCE_PEER_SET` which is used everywhere instead of `sc_peerset::SetId::from(0)`. That would help me understand the context.

tomaka commented:
I have finally made the CI pass. The Substrate integration test was failing because of [this bug](https://github.com/paritytech/substrate/pull/7700/commits/16ace7126c9152a32dd2eed646598a57d9a8850d), while the Polkadot integration test was failing because I didn't understand that collators shouldn't be in the validation peers set.

Will finish addressing review concerns.

A burn-in has been started here: https://github.com/paritytech/polkadot/pull/2098

tomaka commented:
> Testing is showing that, when connecting to Polkadot, we don't receive any notification on sets 1 and 2 (transactions and grandpa).
> 
> I believe that there is a compatibility problem with the fact that we only later add peers to these sets.
> When the remote tries to open their transaction and grandpa substreams, they face a rejection from the local node. Later, when the local node attempts to connect, the remote still remembers this rejection and doesn't try to re-open its own substream.
> 
> I'm pushing a commit that changes the behaviour in order to test this hypothesis.
> EDIT: latest commit ([0e91d3d](https://github.com/paritytech/substrate/commit/0e91d3d5d340058726bf1230e69d755a793fd7cf)) indeed fixes the problem.

To continue on this, this "hack" is necessary to maintain backwards compatibility with nodes that don't have https://github.com/paritytech/substrate/pull/7464, which hasn't been released yet.

Rather than going through complicated means to maintain backwards compatibility in a clean way, I'd suggest keeping this "hack" in until the release of 0.7.27 (which is expected to happen tomorrow) then remove it.
If agreed, I'll open an issue.

tomaka commented:
Burn-in shows that the node seems to work correctly.
In particular, the number of sync peers is higher than before.

Bandwidth has been reduced. Variations in the bandwidth seems to sometimes happen when deploying versions of PRs, and I've never been able to figure out why.

tomaka commented:
bot merge
